frontend/controller/service-worker.js

```diff
@@ -13,23 +15,171 @@ sbp('sbp/selectors/register', {
     // TODO: move ahead with encryption stuff ignoring this service worker stuff for now
     // TODO: improve updating the sw: https://stackoverflow.com/a/49748437
     // NOTE: user should still be able to use app even if all the SW stuff fails
+    if (!('serviceWorker' in navigator)) { return }
+
     try {
-      await navigator.serviceWorker.register('/assets/js/sw-primary.js', {
-        scope: '/'
-      })
-      navigator.serviceWorker.addEventListener('message', function (event) {
-        console.info('Received message from SW:', event.data)
+      const swRegistration = await navigator.serviceWorker.register('/assets/js/sw-primary.js', { scope: '/' })
+
+      if (swRegistration) {
+        swRegistration.active?.postMessage({ type: 'store-client-id' })
+      }
+
+      navigator.serviceWorker.addEventListener('message', event => {
+        console.debug('[sw] Received a message from the service worker :', event)
+        const data = event.data
+
+        if (typeof data === 'object' && data.type) {
+          switch (data.type) {
+            case 'pushsubscriptionchange': {
+              sbp('service-worker/resubscribe-push', data.subscription)
+              break
+            }
+            default:
+              console.error('[sw] Received unknown message type from the service worker:', data)
+              break
+          }
+        }
       })
-      setTimeout(function () {
-        navigator.serviceWorker.controller.postMessage({
-          type: 'set',
-          key: 'foo',
-          value: 'bar'
-        })
-      }, 1000)
     } catch (e) {
-      // TODO: this
       console.error('error setting up service worker:', e)
     }
+  },
+  'service-worker/setup-push-subscription': async function (followingNotification = null) {
```

Comment:
2 questions:

1. Can `'service-worker/setup-push-subscription'` be called safely multiple times? I ask because `requestNotificationPermission`, where it's called, can be called multiple times.
2. `followingNotification` again, should probably be removed. It doesn't seem to be used anyway.

Reply:
@taoeffect 
1. Yes, `service-worker/setup-push-subscription` only takes the full steps of generating a new push subscription and sending/storing it to the server when it's the first time executing it. For any subsequent calls, it only sends the existing subscription to the server to make sure the server keeps the subscription on its side.

2. no prob.